cores/esp32/IPAddress.h

```diff
@@ -1,114 +1,117 @@
 
 public:
     // Constructors
+
+    // Default IPv4
     IPAddress();
     IPAddress(IPType ip_type);
     IPAddress(uint8_t first_octet, uint8_t second_octet, uint8_t third_octet, uint8_t fourth_octet);
-    IPAddress(uint8_t o1, uint8_t o2, uint8_t o3, uint8_t o4, uint8_t o5, uint8_t o6, uint8_t o7, uint8_t o8, uint8_t o9, uint8_t o10, uint8_t o11, uint8_t o12, uint8_t o13, uint8_t o14, uint8_t o15, uint8_t o16);
+    IPAddress(uint8_t o1, uint8_t o2, uint8_t o3, uint8_t o4, uint8_t o5, uint8_t o6, uint8_t o7, uint8_t o8, uint8_t o9, uint8_t o10, uint8_t o11, uint8_t o12, uint8_t o13, uint8_t o14, uint8_t o15, uint8_t o16, uint8_t z=0);
+     // IPv4; see implementation note
     IPAddress(uint32_t address);
+     // Default IPv4
     IPAddress(const uint8_t *address);
-    IPAddress(IPType ip_type, const uint8_t *address);
+    IPAddress(IPType ip_type, const uint8_t *address, uint8_t z=0);
     // If IPv4 fails tries IPv6 see fromString function
     IPAddress(const char *address);
-    virtual ~IPAddress() {}
+    IPAddress(const IPAddress& address);
 
     bool fromString(const char *address);
     bool fromString(const String &address) { return fromString(address.c_str()); }
 
-    // Overloaded cast operator to allow IPAddress objects to be used where a
-    // uint32_t is expected
-    operator uint32_t() const
-    {
-        return _type == IPv4 ? _address.dword[IPADDRESS_V4_DWORD_INDEX] : 0;
-    }
+    // Overloaded cast operator to allow IPAddress objects to be used where a uint32_t is expected
+    // NOTE: IPv4 only; see implementation note
+    operator uint32_t() const { return _type == IPv4 ? _address.dword[IPADDRESS_V4_DWORD_INDEX] : 0; };
 
     bool operator==(const IPAddress& addr) const;
+    bool operator!=(const IPAddress& addr) const { return !(*this == addr); };
+
+    // NOTE: IPv4 only; we don't know the length of the pointer
     bool operator==(const uint8_t* addr) const;
 
     // Overloaded index operator to allow getting and setting individual octets of the address
     uint8_t operator[](int index) const;
     uint8_t& operator[](int index);
 
     // Overloaded copy operators to allow initialisation of IPAddress objects from other types
+    // NOTE: IPv4 only
     IPAddress& operator=(const uint8_t *address);
+    // NOTE: IPv4 only; see implementation note
     IPAddress& operator=(uint32_t address);
     // If IPv4 fails tries IPv6 see fromString function
     IPAddress& operator=(const char *address);
+    IPAddress& operator=(const IPAddress& address);
 
     virtual size_t printTo(Print& p) const;
-    String toString() const;
+    size_t printTo(Print& p, bool includeZone) const;
+    String toString(bool includeZone = false) const;
 
     IPType type() const { return _type; }
 
-    friend class EthernetClass;
+    uint8_t zone() const { return (type() == IPv6)?_zone:0; }
+
+    // LwIP conversions
+    void to_ip_addr_t(ip_addr_t* addr) const;
+    IPAddress& from_ip_addr_t(ip_addr_t* addr);
```

Comment:
Is it possible to pass argument as `const ip_addt_t* addr`:

```
    IPAddress& from_ip_addr_t(const ip_addr_t* addr);
```

Reply:
Sorry for another request, could it be possible to have a constructor that takes directly `const ip_addr_t*`:

`IPAddress(const ip_addr_t*)`

Actually the only use-case I see is for initializing a IPAddress from `ip_addr_t` so it requires `IPAddress().from_ip_addr_t()` which is a small waste.

Reply:
Maybe also add `explicit` to that constructor, just to be sure.